### PR TITLE
[receiver/elasticsearch] Implement Client

### PR DESCRIPTION
**Pull request by BinaryFissionGames**
**Description:**
* Implement client for scraping metrics from an Elasticsearch cluster.

**Testing:**
* Unit tests added for scraper client

### CODE REVIEW

File: receiver/elasticsearchreceiver/testdata/sample_payloads/health.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    "active_shards": 0,
+    "relocating_shards": 0,
+    "initializing_shards": 0,
+    "unassigned_shards": 0,
+    "delayed_unassigned_shards": 0,
+    "number_of_pending_tasks": 0,
+    "number_of_in_flight_fetch": 0,
+    "task_max_waiting_in_queue_millis": 0,
+    "active_shards_percent_as_number": 100.0
+}
```
Comment (djaglowski) [inline receiver/elasticsearchreceiver/testdata/sample_payloads/health.json:17]:
```suggestion
}

```

File: receiver/elasticsearchreceiver/client.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+type defaultElasticsearchClient struct {
+	client     *http.Client
+	endpoint   *url.URL
+	authHeader string
+	logger     *zap.Logger
+}
+
+var _ elasticsearchClient = (*defaultElasticsearchClient)(nil)
+
+func newElasticsearchClient(logger *zap.Logger, client *http.Client, endpoint *url.URL, username, password string) *defaultElasticsearchClient {
```
Comment (djaglowski) [inline receiver/elasticsearchreceiver/client.go:55]: Would it make more sense to just pass the config, use it to create the client here? Seems unusual that we would expect it to be created elsewhere first.

File: receiver/elasticsearchreceiver/client.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	case 200: // OK
+	case 401:
+		return nil, errUnauthenticated
+	case 403:
+		return nil, errUnauthorized
+	default:
+		return nil, fmt.Errorf("got non 200 status code %d", resp.StatusCode)
+	}
+
+	return io.ReadAll(resp.Body)
```
Comment (djaglowski) [inline receiver/elasticsearchreceiver/client.go:152]: The if condition and switch have some redundant checking (though I see why it was convenient to separate them). Still, I have to wonder if there's a better way. What if we bail as soon as we know it's a 200. Then if we're still going it's a non-200.

```suggestion
	if resp.StatusCode == 200 {
		return io.ReadAll(resp.Body)
	}
	
	body, err := io.ReadAll(resp.Body)
	c.logger.Debug(
		"Failed to make request to Elasticsearch",
		zap.String("path", path),
		zap.Int("status_code", resp.StatusCode),
		zap.ByteString("body", body),
		zap.NamedError("body_read_error", err),
	)
	
	switch resp.StatusCode {
	case 401:
		return nil, errUnauthenticated
	case 403:
		return nil, errUnauthorized
	default:
		return nil, fmt.Errorf("got non 200 status code %d", resp.StatusCode)
	}
```

File: receiver/elasticsearchreceiver/internal/model/nodestats.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/elasticsearchreceiver/internal/model"
+
+// NodeStats represents a response from elasticsearch's /_nodes/stats endpoint.
+// The struct is not exhaustive; It does not provide all values returned by elasticsearch,
+// only the ones relevant to the metrics retrieved by the scraper.
+type NodeStats struct {
```
Comment (djaglowski) [inline receiver/elasticsearchreceiver/internal/model/nodestats.go:20]: I appreciate the readability this brings but one concern I always have with this kind of structural representation is that we may end up losing the ability to differentiate between missing values and zero values. Of course, this is only a concern if the API might actually return a result that is missing some of the possible keys we're looking for. Do you know if that is the case here?

Comment (BinaryFissionGames) [reply receiver/elasticsearchreceiver/internal/model/nodestats.go:20]: I don't think there's any case where the response omits keys.

The one exception, which is already being accounted for, are fields marked as "linux only" on the [api docs](https://www.elastic.co/guide/en/elasticsearch/reference/7.16/cluster-nodes-stats.html). The only field we use that's linux only is io_stats, which we have as a pointer here, so we can differentiate between whether it was actually omitted or not.

Comment (djaglowski) [reply receiver/elasticsearchreceiver/internal/model/nodestats.go:20]: Great, thanks for explaining. Looks good then.

**Comment by djaglowski** [comment]
I think this looks good, but please rebase.

**Comment by BinaryFissionGames** [comment]
@djaglowski Rebased; I can see the changelog causing more friction with these PRs as time goes on here, since that seems to change a lot now.

**Comment by djaglowski** [comment]
> @djaglowski Rebased; I can see the changelog causing more friction with these PRs as time goes on here, since that seems to change a lot now.

Yes, #6924 is tracking this concern. Ultimately, there seems to be a tradeoff between developer experience and user experience (from the changelog omitting important changes). Ideally, a solution would be identified that solves for both concerns, but none has been articulated yet.

**Review by djaglowski — APPROVED** [review]

**Pull request merged**